### PR TITLE
feat: Optimize for CASE WHEN .. THEN column ELSE null END

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion-comet/issues/669

## Rationale for this change

`CaseExpr` is an expensive expression and there are some cases where can use specialized expressions instead.

For the usage `CASE WHEN predicate THEN column ELSE null END` which is seen in numerous TPC-DS queries, we can just evaluate the `column` and then change the null mask based on the `predicate`, which is a much faster operation (more than 10x faster than using the standard case expression).

## How are these changes tested?

I manually tested with this query.

```sql
select
    sum(case when (ss_quantity=1) then ss_sales_price else null end) sun_sales,
    sum(case when (ss_quantity=2) then ss_sales_price else null end) mon_sales,
    sum(case when (ss_quantity=3) then ss_sales_price else null end) tue_sales,
    sum(case when (ss_quantity=4) then ss_sales_price else null end) wed_sales,
    sum(case when (ss_quantity=5) then ss_sales_price else null end) thu_sales,
    sum(case when (ss_quantity=6) then ss_sales_price else null end) fri_sales,
    sum(case when (ss_quantity=7) then ss_sales_price else null end) sat_sales
from store_sales;
```

Here are the timings in seconds for five iterations.

### Spark
```
[2.5691187381744385, 1.3875913619995117, 1.2080309391021729, 1.4287035465240479, 1.2525527477264404]
```

### Comet before this PR
```
[4.076025724411011, 2.6956324577331543, 2.5712761878967285, 2.5598058700561523, 2.5548157691955566]
```

### Comet with this PR
```
[3.372154951095581, 1.8503100872039795, 1.7506794929504395, 1.722181797027588, 1.696714162826538]
```

### CODE REVIEW

**Comment by andygrove** [comment]
I also see a ~5% performance improvement with TPC-DS q43 @ sf=100

**Comment by andygrove** [comment]
@parthchandra @kazuyukitanimura @huaxingao @viirya This is ready for review now

File: native/core/src/execution/datafusion/planner.rs [hunk 1]
```diff
@@ -541,6 +542,24 @@ impl PhysicalPlanner {
                         Some(self.create_expr(case_when.else_expr.as_ref().unwrap(), input_schema)?)
                     }
                 };
+
+                // optimized path for CASE WHEN predicate THEN expr ELSE null END
+                if else_phy_expr.is_none() && when_then_pairs.len() == 1 {
```
Comment (parthchandra) [inline native/core/src/execution/datafusion/planner.rs:547]: Couldn't we have added this as an optimization in DF itself?

Comment (andygrove) [reply native/core/src/execution/datafusion/planner.rs:547]: I filed https://github.com/apache/datafusion/issues/11484 to track up streaming this

File: native/spark-expr/src/case_expr_or_null.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        if let ColumnarValue::Array(bit_mask) = self.predicate.evaluate(batch)? {
+            let bit_mask = bit_mask
+                .as_any()
+                .downcast_ref::<BooleanArray>()
+                .expect("predicate should evaluate to a boolean array");
+            // invert the bitmask
+            let bit_mask = arrow::compute::kernels::boolean::not(bit_mask)?;
+            match self.expr.evaluate(batch)? {
+                ColumnarValue::Array(array) => Ok(ColumnarValue::Array(nullif(&array, &bit_mask)?)),
+                ColumnarValue::Scalar(_) => Err(DataFusionError::Execution(
```
Comment (comphead) [inline native/spark-expr/src/case_expr_or_null.rs:108]:
```suggestion
                ColumnarValue::Scalar(_) => exec_err!(
```

File: native/spark-expr/src/case_expr_or_null.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                    "expression did not evaluate to an array".to_string(),
+                )),
+            }
+        } else {
+            Err(DataFusionError::Execution(
```
Comment (comphead) [inline native/spark-expr/src/case_expr_or_null.rs:113]:
```suggestion
            exec_err!
```

File: native/core/src/execution/datafusion/planner.rs [hunk 4]
```diff
@@ -541,6 +542,24 @@ impl PhysicalPlanner {
...
+                    let when_then = &when_then_pairs[0];
+                    // CaseWhenExprOrNull is only safe to use for expressions that do not
+                    // have side effects, and it is only suitable to use for expressions
+                    // that are inexpensive to compute (such as a column reference)
+                    // because it will be evaluated for all rows in the batch rather
+                    // than just the rows where the predicate is true.
+                    // For now, we limit the use to raw column references
+                    if when_then.1.as_any().is::<Column>() {
+                        return Ok(Arc::new(CaseWhenExprOrNull::new(
+                            when_then.0.clone(),
```
Comment (comphead) [inline native/core/src/execution/datafusion/planner.rs:557]:
```suggestion
                            Arc::clone(&when_then.0),
                            Arc::clone(&when_then.1),
```

**Comment by andygrove** [comment]
Let's just wait to get these optimizations from upstream DataFusion. I can test against the lastest DF code when benchmarking.

**Pull request closed without merging**